### ISSUE TITLE
Joint validation raise_for leads to `This is a software defect. Please report!`

### ISSUE THREAD

**Issue by jsheunis** [datalad/datalad-next#358]
I am busy developing some constraint validation code, using `datalad-next`. Following example implementations in `datalad-next`'s codebase, I have the following code

```python
from datalad_catalog.constraints import (
    EnsureWebCatalog,
)
from datalad_catalog.webcatalog import (
    WebCatalog,
)
from datalad_next.commands import (
    EnsureCommandParameterization,
    ParameterConstraintContext,
)

class CreateParameterValidator(EnsureCommandParameterization):
    """"""
    def _check_force(self, catalog, force):
        """"""
        if catalog.is_created() and not force:
            self.raise_for(
                catalog.location,
                'the force flag should be True to overwrite the assets of an existing catalog',
            )

    def __init__(self):
        super().__init__(
            param_constraints=dict(
                catalog=EnsureWebCatalog(),
                metadata=EnsureGeneratorFromFileLike(EnsureJSON()),
                config_file=EnsurePath(lexists=True),
                force=EnsureBool(),
            ),
            joint_constraints={
                ParameterConstraintContext(('catalog', 'force'),
                                            'force-when-existing'):
                    self._check_force,
            },
        )
```

This class is a validator for `Create(ValidatedInterface)`, which is exposed as the `catalog-create` entrypoint. I do a test for the joint validation by first creating a catalog at a path does not previously exist:

```
datalad catalog-create -c ../Data/kaaap2
```

When I run the same command again, without the force flag, the joint validation should raise an error.

However, I get the following:

```
[ERROR  ] on error the joint validator <bound method CreateParameterValidator._check_force of CreateParameterValidator()> did not a mapping of parameter name to (violating) value comprising all constraint context parameters. This is a software defect. Please report!
```

<details>
<summary>Debug output:</summary>

```
datalad --dbg catalog-create -c ../Data/kaaap2

Traceback (most recent call last):
  File "/Users/jsheunis/Documents/psyinf/dlnext/datalad_next/constraints/parameter.py", line 220, in joint_validation
    res = validator(**{p: validated[p] for p in ctx.parameters})
  File "/Users/jsheunis/Documents/psyinf/datalad-catalog/datalad_catalog/create.py", line 55, in _check_force
    self.raise_for(
  File "/Users/jsheunis/Documents/psyinf/dlnext/datalad_next/constraints/base.py", line 52, in raise_for
    raise ConstraintError(self, value, msg)
datalad_next.constraints.exceptions.ConstraintError: the force flag should be True to overwrite the assets of an existing catalog

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/jsheunis/opt/miniconda3/envs/meow3/bin/datalad", line 8, in <module>
    sys.exit(main())
  File "/Users/jsheunis/opt/miniconda3/envs/meow3/lib/python3.10/site-packages/datalad/cli/main.py", line 143, in main
    _run(cmdlineargs)
  File "/Users/jsheunis/opt/miniconda3/envs/meow3/lib/python3.10/site-packages/datalad/cli/main.py", line 167, in _run
    ret = _run_with_debugger(namespace) \
  File "/Users/jsheunis/opt/miniconda3/envs/meow3/lib/python3.10/site-packages/datalad/cli/main.py", line 181, in _run_with_debugger
    return cmdlineargs.func(cmdlineargs)
  File "/Users/jsheunis/opt/miniconda3/envs/meow3/lib/python3.10/site-packages/datalad/cli/exec.py", line 107, in call_from_parser
    ret = list(ret)
  File "/Users/jsheunis/Documents/psyinf/dlnext/datalad_next/patches/interface_utils.py", line 144, in _execute_command_
    allkwargs = param_validator(
  File "/Users/jsheunis/Documents/psyinf/dlnext/datalad_next/constraints/parameter.py", line 370, in __call__
    final = self.joint_validation(validated, on_error)
  File "/Users/jsheunis/Documents/psyinf/dlnext/datalad_next/constraints/parameter.py", line 224, in joint_validation
    raise RuntimeError(
RuntimeError: on error the joint validator <bound method CreateParameterValidator._check_force of CreateParameterValidator()> did not a mapping of parameter name to (violating) value comprising all constraint context parameters. This is a software defect. Please report!
```

</details>

**Comment by mih** [comment]
I have adjusted the error message in #360 to be more precise. This would be a valid `raise_for()` call:

```py
self.raise_for(
    dict(catalog=catalog, force=force),
    'the force flag should be True to overwrite the assets of an existing catalog',
)
```

This was initially not documented well. Fixed a couple of days ago in b1655ccb0b1db8f0e8ddbc6e37e67899aca30696

**Comment by jsheunis** [comment]
Thanks @mih! Confirmed that this fixes my issue.
